### PR TITLE
[SPARK-28382][SQL] Add array function - unnest

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

Function | Return Type | Description | Example | Result
-- | -- | -- | -- | --
unnest(anyarray) | set of  anyelement | expand an array to a set of rows | unnest(ARRAY[1,2]) | 1 2 (2 rows)

https://www.postgresql.org/docs/11/functions-array.html

`unnest` is not `explode`, but like explode recursively

```sql
postgres=# select unnest(null);
ERROR:  function unnest(unknown) is not unique
LINE 1: select unnest(null);
               ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
postgres=# select unnest(array[null]);
 unnest
--------

(1 row)

postgres=# select unnest(array[null, null]);
 unnest
--------


(2 rows)

postgres=# select unnest(array[array[1, 2], array[3, 4]]);
 unnest
--------
      1
      2
      3
      4
(4 rows)

postgres=# select unnest(array[array[1, 2], array[3, 4, 5]]);
ERROR:  multidimensional arrays must have array expressions with matching dimensions
```

### Why are the changes needed?

from [SPARK-27764: Feature Parity between PostgreSQL and Spark](https://issues.apache.org/jira/browse/SPARK-27764)

### Does this PR introduce any user-facing change?

add an array generator.

### How was this patch tested?

add uts

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @wangyum thanks in advance

**Comment by maropu** [comment]
This is a general question; is this function common in the other DBMS-like systems? Also, is there the array pattern that `unnest` can't expand and `explode` can?  If no, we don't need `explode` any more when `unnest` implemented?

**Comment by yaooqinn** [comment]
@maropu explode support map type

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 1]
```diff
@@ -343,6 +343,63 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+       10
+       20
+      > SELECT _FUNC_(a) FROM VALUES (array(1,2)), (array(3,4)) AS v1(a);
+       1
+       2
+       3
+       4
+  """,
+  since = "3.0.0")
+case class UnNest(child: Expression) extends UnaryExpression with Generator with CodegenFallback {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Why can't we reuse `explode` here?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: forgive me for not getting your question’s point here, do you mean they are equivalent?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 2]
```diff
@@ -343,6 +343,63 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(10, 20));
+       10
+       20
+      > SELECT _FUNC_(a) FROM VALUES (array(1,2)), (array(3,4)) AS v1(a);
+       1
+       2
+       3
+       4
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: @yaooqinn given this examples, yes looks equivalent. How are they different in this PR?

```scala
scala> sql("select unnest(a) FROM VALUES (array(1,2)), (array(3,4)) AS v1(a)").show()
+------+
|unnest|
+------+
|     1|
|     2|
|     3|
|     4|
+------+


scala> sql("select unnest(array(10, 20))").show()
+------+
|unnest|
+------+
|    10|
|    20|
+------+


scala> sql("select explode(a) FROM VALUES (array(1,2)), (array(3,4)) AS v1(a)").show()
+---+
|col|
+---+
|  1|
|  2|
|  3|
|  4|
+---+


scala> sql("select explode(array(10, 20))").show()
+---+
|col|
+---+
| 10|
| 20|
+---+
```

**Comment by HyukjinKwon** [comment]
```sql
postgres=# select unnest(array[array[1, 2], array[3, 4]]);
 unnest
--------
      1
      2
      3
      4
(4 rows)

```

this case seems unsupported in this PR:

```scala
scala> sql("select unnest(array(array(1, 2), array(1, 2)))").show()
+------+
|unnest|
+------+
+------+
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Sorry，the examples are misleading，should use multidimensional arrays i.e array(array(1,2),array(3,4)). Then it will show the difference between unnest and explode. I
 will add some new examples here.

**Comment by yaooqinn** [comment]
I will do re-check, thanks very much @HyukjinKwon

**Comment by yaooqinn** [comment]
> ```sql
> postgres=# select unnest(array[array[1, 2], array[3, 4]]);
>  unnest
> --------
>       1
>       2
>       3
>       4
> (4 rows)
> ```
> 
> this case seems unsupported in this PR:
> 
> ```scala
> scala> sql("select unnest(array(array(1, 2), array(1, 2)))").show()
> +------+
> |unnest|
> +------+
> +------+
> ```

these case should be supported now with the lastest commit

File: sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala [hunk 3]
```diff
@@ -344,6 +344,7 @@ class GeneratorFunctionSuite extends QueryTest with SharedSparkSession {
   }
 }
 
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala]: I would remove unnecessary changes.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql [hunk 4]
```diff
@@ -1089,10 +1089,9 @@ order by fault;
 -- left join unnest(v1ys) as u1(u1y) on u1y = v2y;
 
 -- [SPARK-28382] Array Functions: unnest
--- select * from
--- (values (1, array(10,20)), (2, array(20,30))) as v1(v1x,v1ys)
--- left join (values (1, 10), (2, 20)) as v2(v2x,v2y) on v2x = v1x
--- left join unnest(v1ys) as u1(u1y) on u1y = v2y;
+create or replace temporary view v1 as select * from values (1, array(10,20)), (2, array(20,30)) as v1(v1x,v1ys);
+create or replace temporary view v2 as select * from values (1, 10), (2, 20) as v2(v2x,v2y);
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql]: Hm, why is it different from commented? Does it match to PostgreSQL's?

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/aggregates_part4.sql [hunk 5]
```diff
@@ -404,12 +403,8 @@
...
 -- [SPARK-28382] Array Functions: unnest
 -- check collation-sensitive matching between grouping expressions
--- select v||'a', case v||'a' when 'aa' then 1 else 0 end, count(*)
---   from unnest(array['a','b']) u(v)
---  group by v||'a' order by 1;
--- select v||'a', case when v||'a' = 'aa' then 1 else 0 end, count(*)
---   from unnest(array['a','b']) u(v)
---  group by v||'a' order by 1;
+select v||'a', case v||'a' when 'aa' then 1 else 0 end, count(*) from (select unnest(array('a','b')) as v) u group by v||'a' order by 1;
+select v||'a', case when v||'a' = 'aa' then 1 else 0 end, count(*) from (select unnest(array('a','b')) as v) u group by v||'a' order by 1;
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/aggregates_part4.sql]: Why should we inline them? can't we just uncomment the tests?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 6]
```diff
@@ -343,6 +344,87 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+      > SELECT _FUNC_(a) FROM VALUES (array(array(1,2), array(3,4))) AS v1(a);
+       1
+       2
+       3
+       4
+  """,
+  since = "3.0.0")
+case class UnNest(child: Expression) extends UnaryExpression with Generator with CodegenFallback {
+
+  override def prettyName: String = "unnest"
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: I think we don't need to override this since it's matched to class name

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 7]
```diff
@@ -343,6 +344,87 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+
+  override def elementSchema: StructType = {
+    new StructType().add(prettyName, getLeafDataType(child.dataType), true)
+  }
+
+  // TODO: multidimensional arrays must have array expressions with matching dimensions
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: what does this mean?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 8]
```diff
@@ -343,6 +344,87 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+       3
+       4
+      > SELECT _FUNC_(a) FROM VALUES (array(array(1,2), array(3,4))) AS v1(a);
+       1
+       2
+       3
+       4
+  """,
+  since = "3.0.0")
+case class UnNest(child: Expression) extends UnaryExpression with Generator with CodegenFallback {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Can we implement codegen as well?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala [hunk 9]
```diff
@@ -343,6 +344,87 @@ abstract class ExplodeBase extends UnaryExpression with CollectionGenerator with
...
+    var inputArray = child.eval(input).asInstanceOf[ArrayData]
+    if (inputArray == null) {
+      Nil
+    } else {
+      val rows = ArrayBuffer[InternalRow]()
+      var currentType = child.dataType
+      while (currentType.isInstanceOf[ArrayType]) {
+        val currentElementType = getElementTypeOrItself(currentType)
+        if (!currentElementType.isInstanceOf[ArrayType]) {
+          inputArray.foreach(currentElementType, (_, e) => rows += InternalRow(e))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Can we avoid Scala's foreach in such critical path (see https://github.com/databricks/scala-style-guide#traversal-and-zipwithindex)?  We can at least switch to a regular for loop instead.

File: sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala]: thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: pg can check nested arrays for each element's dimension, we seem not able to do that

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 4 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql]: we don't support unnest as table here yet

Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql]: I can revert these changes if we shouldn't do that

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 9 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: is this a scala foreach? I guess it is a Spark ArrayData' own implementation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 8 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: seems very difficult to do codegen

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 4 above)
Comment (HyukjinKwon) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql]: Let's revert. I think the point of point PostgreSQL tests is to support the same syntax.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 7 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Sorry I wasn't clear. what does "nested arrays for each element's dimension" mean?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 8 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: Hm, why? similar expressions look being able to do it.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: for each round of unnesting, we have to know what dataType it is at that time.

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql (hunk 4 above)
Comment (yaooqinn) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/join.sql]: ok

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala]: array(array(1,2), array(3,4)) √
array(array(1,2), array(3,4,5)) ×

**Comment by HyukjinKwon** [comment]
Thanks @yaooqinn. Will take a closer look soon within few days.

**Comment by yaooqinn** [comment]
> Thanks @yaooqinn. Will take a closer look soon within few days.

You're welcome. And thanks very much for your patience and help.

**Comment by yaooqinn** [comment]
retest this please

**Review by HyukjinKwon — COMMENTED** [review]
Since we're [re-considering about PostgreSQL dialect due to large differences between them](http://apache-spark-developers-list.1001551.n3.nabble.com/DISCUSS-PostgreSQL-dialect-td28417.html#a28421), can you check if there are other DBMSes as other references?

Implementation itself seems fine but now seems we should have a better reason than PostgreSQL

**Comment by yaooqinn** [comment]
close this because it seems pg only,  and other systems have their own different implementations

**Comment by yaooqinn** [comment]
close this because it seems pg only,  and other systems have their own different implementations
